hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestMetaRegionReplicaReplicationEndpoint.java

```diff
@@ -225,6 +227,38 @@ public void testCatalogReplicaReplicationWithFlushAndCompaction() throws Excepti
     }
   }
 
+  @Test
+  public void testCatalogReplicaReplicationWALRolledAndDeleted() throws Exception {
+    Connection connection = ConnectionFactory.createConnection(HTU.getConfiguration());
+    TableName tableName = TableName.valueOf("hbase:meta");
+    Table table = connection.getTable(tableName);
+    try {
+      MiniHBaseCluster cluster = HTU.getHBaseCluster();
+      HRegionServer hrs = cluster.getRegionServer(cluster.getServerHoldingMeta());
+      ReplicationSource source = (ReplicationSource) hrs.getReplicationSourceService()
+        .getReplicationManager().catalogReplicationSource.get();
+      ((ReplicationPeerImpl) source.replicationPeer).setPeerState(false);
+      // load the data to the table
+      for (int i = 0; i < 5; i++) {
+        LOG.info("Writing data from " + i * 1000 + " to " + (i * 1000 + 1000));
+        HTU.loadNumericRows(table, HConstants.CATALOG_FAMILY, i * 1000, i * 1000 + 1000);
+        LOG.info("flushing table");
+        HTU.flush(tableName);
+        LOG.info("compacting table");
+        if (i < 4) {
+          HTU.compact(tableName, false);
+        }
+      }
+      HTU.getHBaseCluster().getMaster().getLogCleaner().triggerCleanerNow().get(1,
+        TimeUnit.SECONDS);
+      ((ReplicationPeerImpl) source.replicationPeer).setPeerState(true);
+      verifyReplication(tableName, numOfMetaReplica, 0, 5000, HConstants.CATALOG_FAMILY);
```

Comment:
Here we just checked whether data can be read? I think the main thing here is that we should make sure the ReplicationSource can still relicate things out, i.e, it is not stuck forever.

So maybe we should load some data, flush, delete the WAL, enable the replication source, and then load more data, without flushing, to make sure that we could read all the data?

Reply:
>So here, since the data have already been flushed out, even if the replication is stuck forever, it is still possible that we could read the data from the secondary replica, as all the data are in hfiles now.

Not really. The secondary replicas only know they need to load new files when they get notified that flush/compaction happened in the primary. This is done via replication, though the WAL markers. With replication stuck, the secondaries never load new files, it only gets worse, as compacted files eventually get deleted by the compaction discharger, secondaries now start to throw FNFE. This is the scenario this test is simulating.

Reply:
Yes, reassignment would "fix" the issue. But we are not triggering reassignment in this test.

Reply:
But we do not disable balancer in this test right? And maybe we will add some periodically reloading logic in the future which could make the test not test what we want. So I think still think, we should try to load more data, where the new data are only in WAL file, to make sure that the secondary replica can still load them.

Reply:
Ok, changed it to disable balancer and also add extra data after peer is reenabled but no flush/compaction triggered.